File: py/configu/stores/json_file_store.py
import json
from typing import List

from ..core import Config, ConfigStore, ConfigStoreQuery


class JsonFileConfigStore(ConfigStore):
    """A `ConfigStore` persisted in a json file"""

    _path: str

    def __init__(self, path: str) -> None:
        super().__init__(type="json-file")
        self._path = path

    def _read(self) -> List[Config]:
        """
        Reads JSON file contents
        :return: The file contents, represented as a list of Config
        dicts
        """
        with open(self._path, mode="r", encoding="utf-8") as json_file:
            data = json.load(json_file)
        return [Config(**args) for args in data]

    def _write(self, configs: List[Config]) -> None:
        """
        Writes to the JSON file
        :param configs: a list of key/set/value dicts to write
        """
        data = [config.to_dict() for config in configs]
        with open(self._path, mode="w", encoding="utf-8") as json_file:
            json.dump(data, json_file)

    def get(self, queries: List[ConfigStoreQuery]) -> List[Config]:
        stored_configs = self._read()
        query_ids = [f"{query.set}.{query.key}" for query in queries]
        return [
            config
            for config in stored_configs
            if f"{config.set}.{config.key}" in query_ids
        ]

    def set(self, configs: List[Config]) -> None:
        stored_configs = self._read()
        set_config_ids = [f"{config.set}.{config.key}" for config in configs]
        existing = [
            config
            for config in stored_configs
            if f"{config.set}.{config.key}" not in set_config_ids
        ]
        next_configs = [config for config in existing + configs if bool(config.value)]
        self._write(next_configs)

File: go/stores/json_file_store.go
package stores

import (
	"encoding/json"
	"fmt"
	"os"

	"github.com/configu/configu/go/core/generated"

	"golang.org/x/exp/slices"
)

type JsonFileConfigStore struct {
	Path string
}

func (s JsonFileConfigStore) Read() []generated.ConfigStoreContentsElement {
	content, _ := os.ReadFile(s.Path)
	res := make([]generated.ConfigStoreContentsElement, 0)
	json.Unmarshal(content, &res)
	return res
}

func (s JsonFileConfigStore) Write(elements []generated.ConfigStoreContentsElement) {
	data, _ := json.Marshal(elements)
	os.WriteFile(s.Path, data, 0644)
}

func (s JsonFileConfigStore) Get(queries []generated.ConfigStoreQuery) []generated.Config {
	stored_configs := s.Read()
	query_ids := make([]string, len(queries))
	for i, query := range queries {
		query_ids[i] = fmt.Sprintf("%s.%s", query.Set, query.Key)
	}
	results := make([]generated.Config, 0)
	for _, d := range stored_configs {
		if slices.Contains(query_ids, fmt.Sprintf("%s.%s", d.Set, d.Key)) {
			results = append(results, generated.Config(d))
		}
	}
	return results
}

func (s JsonFileConfigStore) Set(configs []generated.Config) {
	stored_configs := s.Read()
	set_config_ids := make([]string, len(configs))
	for i, d := range stored_configs {
		set_config_ids[i] = fmt.Sprintf("%s.%s", d.Set, d.Key)
	}
	existing := make([]generated.ConfigStoreContentsElement, 0)
	for _, d := range stored_configs {
		if !slices.Contains(set_config_ids, fmt.Sprintf("%s.%s", d.Set, d.Key)) {
			existing = append(existing, d)
		}
	}
	config_store_elements := make([]generated.ConfigStoreContentsElement, 0)
	for _, d := range configs {
		config_store_elements = append(config_store_elements, generated.ConfigStoreContentsElement(d))
	}
	s.Write(append(existing, config_store_elements...))
}

func (s JsonFileConfigStore) GetType() string {
	return "json"
}
